### PR TITLE
[Fix] fix the bug in Recorders when nn.Module contains the 'inplace' …

### PR DESCRIPTION
**Pull request by cape-zck**
…attribute

## Motivation

Find all nn.Modules in the model that contain the 'inplace'  attribute and set them to False in order to prevent occur error in Recorders using recursion algorithm.

## Modification

Write a new function, it will disassemble the Args architecture .If type 'nn.Module' is detected, determine if it contains an 'inplace' attribute and set False if it does. If none, get the OrderedDict and then iterate through the dictionary to continue the recursive search.
And call this function in class BaseAlgorithm init and class SingleTeacherDistill init function.
Because my academic level is limited and this is my first pulling request, the implementation method and quality may be poor. I sincerely hope that I can get some suggestions from you.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Review by wilxy — COMMENTED** [review]
Thanks for your constructive PR, it can help `MMRazor` improve the robustness on different algorithms and models.
Please make further modification based on my comment to merge this PR into `MMRazor`.

File: mmrazor/models/algorithms/base.py [hunk 1]
```diff
@@ -79,6 +79,10 @@ def __init__(self,
         # Cannot assign module before Module.__init__()
         self.architecture = architecture
 
+        # Find all nn.Modules in the model that contain the 'inplace' attribute
+        # and set them to False
+        self.set_module_inplace_false(architecture, 'self.architecture')
```
[the remark this answers is not in the document]
Comment (cape-zck) [reply mmrazor/models/algorithms/base.py]: Thanks for your suggestions. I have already made modification. In`BaseAlgorithm`class and `SingleTeacherDistill` class add the optional`set_module_inplace: bool = True` and `set_teacher_module_inplace: bool = True`.

File: mmrazor/models/algorithms/base.py [hunk 2]
```diff
@@ -39,6 +39,8 @@ class BaseAlgorithm(BaseModel):
             config of :class:`BaseDataPreprocessor`. Defaults to None.
         init_cfg (dict): The weight initialized config for
             :class:`BaseModule`.
+        module_inplace(bool): Whether to allow student module inplace
```
Comment (wilxy) [inline mmrazor/models/algorithms/base.py:42]: This is the `base class` of all algorithms. Some algorithms (like NAS and Pruning) do not contain `student module`, so the docstring here need to be modified.

**Review by wilxy — APPROVED** [review]
I think this PR solves the problem of the failure on distilling `BN` module, and approve merging these changes.

File: mmrazor/models/algorithms/base.py (hunk 2 above)
Comment (cape-zck) [reply mmrazor/models/algorithms/base.py]: I'm sorry for ignoring this problem. Now I have modified the document. Thank you very much!

**Review by pppppM — APPROVED** [review]

**Pull request merged**